sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -305,6 +306,63 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
     newChildren: IndexedSeq[Expression]): HashPartitioning = copy(expressions = newChildren)
 }
 
+/**
+ * Represents a partitioning where rows are split across partitions based on transforms defined
+ * by `expressions`. `partitionValuesOpt`, if defined, should contain value of partition key(s) in
+ * ascending order, after evaluated by the transforms in `expressions`, for each input partition.
+ * In addition, its length must be the same as the number of input partitions (and thus is a 1-1
+ * mapping), and each row in `partitionValuesOpt` must be unique.
+ *
+ * For example, if `expressions` is `[years(ts_col)]`, then a valid value of `partitionValuesOpt` is
+ * `[0, 1, 2]`, which represents 3 input partitions with distinct partition values. All rows
+ * in each partition have the same value for column `ts_col` (which is of timestamp type), after
+ * being applied by the `years` transform.
+ *
+ * On the other hand, `[0, 0, 1]` is not a valid value for `partitionValuesOpt` since `0` is
+ * duplicated twice.
+ *
+ * @param expressions partition expressions for the partitioning.
+ * @param numPartitions the number of partitions
+ * @param partitionValuesOpt if set, the values for the cluster keys of the distribution, must be
+ *                           in ascending order.
+ */
+case class KeyGroupedPartitioning(
+    expressions: Seq[Expression],
+    numPartitions: Int,
+    partitionValuesOpt: Option[Seq[InternalRow]] = None) extends Partitioning {
+
+  override def satisfies0(required: Distribution): Boolean = {
+    super.satisfies0(required) || {
+      required match {
+        case c @ ClusteredDistribution(requiredClustering, requireAllClusterKeys, _) =>
+          if (requireAllClusterKeys) {
+            // Checks whether this partitioning is partitioned on exactly same clustering keys of
+            // `ClusteredDistribution`.
+            c.areAllClusterKeysMatched(expressions)
+          } else {
+            // We'll need to find leaf attributes from the partition expressions first.
+            val attributes = expressions.flatMap(_.collectLeaves())
+            attributes.forall(x => requiredClustering.exists(_.semanticEquals(x)))
```

Comment:
We need to add some comments to explain the theory here. If the required distribution is `[a, b, c]`, and the key-grouped partitioning is `[f1(a, b), f2(b, c)]`, can we satisfy the distribution requirement?

I think the theory is, as long as the transform function is deterministic, if two rows have the same value of `[a, b, c]`, the calculated key values `[f1(a, b), f2(b, c)]` are also the same.

Reply:
For simplicity this PR only support transforms with a single argument, so the check here is very similar to `HashPartitioning`. I plan to work on the support of multi-arguments as a separate PR.

> I think the theory is, as long as the transform function is deterministic, if two rows have the same value of [a, b, c], the calculated key values [f1(a, b), f2(b, c)] are also the same.

Yea agree with this. It may be more complicated if you have duplicated keys in distribution or/and partitioning though. We also need to think how to implement `EnsureRequirements.reorderJoinPredicates` for this case.

Reply:
I don't think we need to consider `EnsureRequirements.reorderJoinPredicates` here. Whatever the distribution keys are, `[a, b, c]` or `[c, b, a]`, `[f1(a, b), f2(b, c)]` can always satisfy it.